model/common/src/icon4py/model/common/interpolation/rbf_interpolation.py

```diff
@@ -0,0 +1,520 @@
+from types import MappingProxyType, ModuleType
+
+import gt4py.next as gtx
+import numpy as np
+import scipy.linalg as sla
+
+from icon4py.model.common import (
+    dimension as dims,
+    field_type_aliases as fa,
+    type_alias as ta,
+)
+from icon4py.model.common.grid import base as base_grid
+from icon4py.model.common.math.helpers import (
+    cartesian_coordinates_from_zonal_and_meridional_components_on_cells,
+    cartesian_coordinates_from_zonal_and_meridional_components_on_edges,
+    cartesian_coordinates_from_zonal_and_meridional_components_on_vertices,
+)
+from icon4py.model.common.utils import data_allocation as data_alloc
+
+
+class RBFDimension(enum.Enum):
+    CELL = "cell"
+    EDGE = "edge"
+    VERTEX = "vertex"
+    GRADIENT = "midpoint_gradient"
+
+
+RBF_STENCIL_SIZE: dict[RBFDimension, int] = {
+    RBFDimension.CELL: 9,
+    RBFDimension.EDGE: 4,
+    RBFDimension.VERTEX: 6,
+    RBFDimension.GRADIENT: 10,
+}
+
+
+class InterpolationKernel(enum.Enum):
+    GAUSSIAN = 1
+    INVERSE_MULTIQUADRATIC = 3
+
+
+# TODO: Use these as default?
+@dataclasses.dataclass(frozen=True)
+class InterpolationConfig:
+    # for nested setup this value is a vector of size num_domains
+    # and the default value is resolution dependent, according to the namelist
+    # documentation in ICON
+    rbf_vector_scale: dict[RBFDimension, ta.wpfloat] = MappingProxyType(
+        {
+            RBFDimension.CELL: 1.0,
+            RBFDimension.EDGE: 1.0,
+            RBFDimension.VERTEX: 1.0,
+            RBFDimension.GRADIENT: 1.0,
+        }
+    )
+
+    rbf_kernel: dict[RBFDimension, InterpolationKernel] = MappingProxyType(
+        {
+            RBFDimension.CELL: InterpolationKernel.GAUSSIAN,
+            RBFDimension.EDGE: InterpolationKernel.INVERSE_MULTIQUADRATIC,
+            RBFDimension.VERTEX: InterpolationKernel.GAUSSIAN,
+            RBFDimension.GRADIENT: InterpolationKernel.GAUSSIAN,
+        }
+    )
+
+
+def compute_rbf_scale(mean_characteristic_length: ta.wpfloat, dim: RBFDimension):
+    """Compute the default RBF scale factor. This assumes that the Gaussian
+    kernel is used for vertices and cells, and that the inverse multiquadratic
+    kernel is used for edges."""
+
+    threshold = 2.5 if dim == RBFDimension.CELL else 2.0
+    c1 = 0.4 if dim == RBFDimension.EDGE else 1.8
+    if dim == RBFDimension.CELL:
+        c2 = 3.75
+        c3 = 0.9
+    elif dim == RBFDimension.VERTEX:
+        c2 = 3.0
+        c3 = 0.96
+    else:
+        c2 = 2.0
+        c3 = 0.325
+
+    resol = mean_characteristic_length / 1000.0
+    scale = 0.5 / (1.0 + c1 * math.log(threshold / resol) ** c2) if resol < threshold else 0.5
+    return scale * (resol / 0.125) ** c3 if resol <= 0.125 else scale
+
+
+def construct_rbf_matrix_offsets_tables_for_cells(
+    grid: base_grid.BaseGrid,
+) -> data_alloc.NDArray:
+    """Compute the neighbor tables for the cell RBF matrix: rbf_vec_index_c"""
+    c2e2c = grid.connectivities[dims.C2E2CDim]
```

Comment:
I think this connectivity is available on the ICON grid: `icon_grid.connectivities[dims.C2E2C2EDim]`. There is afunction in `grid_manager.py` called `_construct_triangle_edges` that does this.